service/history/workflow/mutable_state_impl.go

```diff
@@ -6252,3 +6260,341 @@ func (ms *MutableStateImpl) RefreshExpirationTimeoutTask(ctx context.Context) er
+	updatedUpdateInfos map[string]*persistencespb.UpdateInfo,
+) {
+	for updateID, ui := range updatedUpdateInfos {
+		if existing, ok := ms.executionInfo.UpdateInfos[updateID]; ok {
+			if CompareVersionedTransition(existing.GetLastUpdateVersionedTransition(), ui.GetLastUpdateVersionedTransition()) == 0 {
+				continue
+			}
+			ms.approximateSize -= existing.Size() + len(updateID)
+		} else {
+			ms.executionInfo.UpdateCount++
+		}
+		ms.executionInfo.UpdateInfos[updateID] = ui
+		ms.approximateSize += ui.Size() + len(updateID)
+	}
+}
+
+func (ms *MutableStateImpl) syncExecutionInfo(current *persistencespb.WorkflowExecutionInfo, incoming *persistencespb.WorkflowExecutionInfo) error {
+	excludedFields := []string{
+		"WorkflowTaskVersion",
+		"WorkflowTaskScheduledEventId",
+		"WorkflowTaskStartedEventId",
+		"WorkflowTaskRequestId",
+		"WorkflowTaskTimeout",
+		"WorkflowTaskAttempt",
+		"WorkflowTaskStartedTime",
+		"WorkflowTaskScheduledTime",
+		"WorkflowTaskOriginalScheduledTime",
+		"WorkflowTaskType",
+		"WorkflowTaskSuggestContinueAsNew",
+		"WorkflowTaskHistorySizeBytes",
+		"WorkflowTaskBuildId",
+		"WorkflowTaskBuildIdRedirectCounter",
+		"VersionHistories",
+		"ExecutionStats",
+		"LastFirstEventTxnId",
+		"ParentClock",
+		"CloseTransferTaskId",
+		"CloseVisibilityTaskId",
+		"CloseVisibilityTaskCompleted",
+		"WorkflowExecutionTimerTaskStatus",
+		"SubStateMachinesByType",
+		"StateMachineTimers",
+		"TaskGenerationShardClockTimestamp",
+	}
+	err := common.MergeProtoExcludingFields(current, incoming, excludedFields...)
+	if err != nil {
+		return err
+	}
+
+	ms.ClearStickyTaskQueue()
+
+	if ms.executionInfo.WorkflowTaskType == enumsspb.WORKFLOW_TASK_TYPE_SPECULATIVE {
+		ms.workflowTaskManager.deleteWorkflowTask()
+	} else {
+		taskInfo := &WorkflowTaskInfo{
+			Version:             incoming.WorkflowTaskVersion,
+			ScheduledEventID:    incoming.WorkflowTaskScheduledEventId,
+			StartedEventID:      incoming.WorkflowTaskStartedEventId,
+			RequestID:           incoming.WorkflowTaskRequestId,
+			WorkflowTaskTimeout: incoming.WorkflowTaskTimeout.AsDuration(),
+			Attempt:             incoming.WorkflowTaskAttempt,
+			StartedTime:         incoming.WorkflowTaskStartedTime.AsTime(),
+			ScheduledTime:       incoming.WorkflowTaskScheduledTime.AsTime(),
+
+			OriginalScheduledTime: incoming.WorkflowTaskOriginalScheduledTime.AsTime(),
+			Type:                  incoming.WorkflowTaskType,
+
+			SuggestContinueAsNew:   incoming.WorkflowTaskSuggestContinueAsNew,
+			HistorySizeBytes:       incoming.WorkflowTaskHistorySizeBytes,
+			BuildId:                incoming.WorkflowTaskBuildId,
+			BuildIdRedirectCounter: incoming.WorkflowTaskBuildIdRedirectCounter,
+		}
+		ms.workflowTaskManager.UpdateWorkflowTask(taskInfo)
+	}
+
+	return nil
+}
+
+func (ms *MutableStateImpl) syncSubStateMachinesByType(incoming map[string]*persistencespb.StateMachineMap) {
+	ms.executionInfo.SubStateMachinesByType = incoming
```

Comment:
hmm I don't think this is enough. `TransitionCount` field for each node is cluster specific.

Reply:
updated using similar logic as `syncHSMNode`